File: 0001-two-sum/0001-two-sum.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
   vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mp;

        for (int i = 0; i < nums.size(); i++) {
            int current = nums[i];
            int required = target - current;

            if (mp.find(required) != mp.end()) {
                return { mp[required], i };
            }

            mp[current] = i;
        }

        return {};
   }
};

File: 0001-two-sum/0001-two-sum.java
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>(); // <Number, Index>

        for (int i = 0; i < nums.length; i++) {
            int moreNeeded = target - nums[i];

            // 1. Agar partner map me pehle se hai, seedha dono indices return karo
            if (map.containsKey(moreNeeded)) {
                return new int[] { map.get(moreNeeded), i };
            }

            // 2. Partner nahi mila, toh apna number aur index map me daal do
            map.put(nums[i], i);
        }

        return new int[] {}; // Khali array agar koi pair na mile
    }
}